main/core.js

```diff
@@ -0,0 +1,174 @@
+'use strict'
+
+const { app, BrowserWindow, dialog } = require('electron')
+const { join } = require('node:path')
+const execa = require('execa')
+const wallet = require('./wallet')
+const assert = require('node:assert')
+const fs = require('node:fs/promises')
+const JSONStream = require('jsonstream')
+const Sentry = require('@sentry/node')
+
+/** @typedef {import('./typings').Context} Context */
+/** @typedef {import('./typings').CoreEvent} CoreEvent */
+/** @typedef {import('./typings').Activity} Activity */
+
+const corePath = join(
+  __dirname,
+  '..',
+  'node_modules',
+  '@filecoin-station',
+  'core',
+  'bin',
+  'station.js'
+)
+console.log('Core binary: %s', corePath)
+
+let online = false
+
+async function setup (/** @type {Context} */ ctx) {
+  ctx.saveModuleLogsAs = async () => {
+    const opts = { defaultPath: 'station.txt' }
+    const win = BrowserWindow.getFocusedWindow()
+    const { filePath } = win
+      ? await dialog.showSaveDialog(win, opts)
+      : await dialog.showSaveDialog(opts)
+    if (filePath) {
+      await fs.writeFile(
+        filePath,
+        (await execa(corePath, ['logs'])).stdout
+      )
+    }
+  }
+  await start(ctx)
+}
+
+async function start (/** @type {Context} */ ctx) {
+  assert(wallet.getAddress(), 'Core requires FIL address')
+  console.log('Starting Core...')
+
+  startEventsProcess(ctx)
+
+  const coreChildProcess = execa(corePath, [], {
+    env: {
+      FIL_WALLET_ADDRESS: wallet.getAddress()
+    }
+  })
+
+  /** @type {string | null} */
+  let coreExitReason = null
+
+  coreChildProcess.on('close', code => {
+    console.log(`Core closed all stdio with code ${code ?? '<no code>'}`)
+
+    ;(async () => {
+      const log = await getLog()
+      Sentry.captureException('Core exited', scope => {
+        // Sentry UI can't show the full 100 lines
+        scope.setExtra('logs', log.split('\n').slice(-10).join('\n'))
+        scope.setExtra('reason', coreExitReason)
+        return scope
+      })
+    })()
+  })
+
+  coreChildProcess.on('exit', (code, signal) => {
+    const reason = signal ? `via signal ${signal}` : `with code: ${code}`
+    const msg = `Core exited ${reason}`
+    console.log(msg)
+    coreExitReason = signal || code ? reason : null
+  })
+}
+
+function startEventsProcess (/** @type {Context} */ ctx) {
+  console.log('Starting Core Events...')
+  const events = execa(corePath, ['events'])
+  assert(events.stdout, 'Events child process has no stdout')
+  events.stdout
+    .pipe(JSONStream.parse(true))
+    .on('data', (/** @type {CoreEvent} */ event) => {
+      switch (event.type) {
+        case 'jobs-completed': {
+          ctx.setTotalJobsCompleted(event.total)
+          break
+        }
+        case 'activity:info': {
+          ctx.recordActivity({
+            source: event.module,
+            type: 'info',
+            message: event.message,
+            date: new Date(event.date).getTime()
```

Comment:
> Because if we receive older events created before the restart, we will attach incorrect timestamps to them.

We're not attaching new timestamps here, the timestamps are already part of the message from `core`. In this line, the timestamp is just parsed and formatted. Does that make sense?

Good that you brought this up though, I'll double check restarting `station events` doesn't lead to duplicate events

Reply:
Ah, no this is no problem since `station events` only streams new events, and doesn't serve any historic content.

In contrast to what I said in the call, this PR actually only shows activity starting from when you started Station.

Reply:
> We're not attaching new timestamps here, the timestamps are already part of the message from core. In this line, the timestamp is just parsed and formatted. Does that resolve your concern?

Yes, thank you for explaining. I might misunderstand what this line does.

> In contrast to what I said in the call, this PR actually only shows activity starting from when you started Station.

Interesting. I would prefer to show all activities, including entries about shutdown/restart. However, it's not critical. We can defer the fix until later if you like.